sound/soc/sof/pcm.c

```diff
@@ -232,8 +233,18 @@ static int sof_pcm_prepare(struct snd_soc_component *component,
 	if (!spcm)
 		return -EINVAL;
 
-	if (spcm->prepared[substream->stream])
-		return 0;
+	if (spcm->prepared[substream->stream]) {
+		if (!spcm->delayed_stop[substream->stream])
+			return 0;
+
+		/*
+		 * this case should be reached in case of xruns where we absolutely
+		 * want to free-up and reset all PCM/DMA resources
+		 */
+		ret = sof_pcm_stream_free(sdev, substream, spcm, substream->stream, true);
```

Comment:
The 'true' argument is problematic for SoundWire devices. As pointed out by @ranj063, in case of an xrun w/ prepare, the alh_id is NOT provided again to the SOF driver so the widget_setup fails.

The solution is to avoid freeing the widgets, which are re-allocated immediately anyways so that's an unnecessary overkill even outside of the SoundWire issue.